File: 13. Graph/11. Articulation Points/Main.java
import java.util.*;
import java.lang.*;
import java.io.*;
import java.math.*;

class Graph {

    int n;
    ArrayList<ArrayList<Integer>> graph;
    static int time = 0;

    Graph (int n) {
        this.n = n;
        this.graph = new ArrayList<>();
        for (int i=0; i<n; i++) {
            graph.add(new ArrayList<Integer>());
        }
    }

    void addEdge (int a, int b) {
        graph.get(a).add(b);
        graph.get(b).add(a);
    }

    void artnPoints () {
        int[] parent = new int[n];
        Arrays.fill(parent, -1);
        boolean[] ap = new boolean[n];
        int[] disc = new int[n];
        Arrays.fill(disc, -1);
        int[] low = new int[n];
        for (int i=0; i<n; i++) {
            if (disc[i] == -1) {
                dfs(i, disc, low, parent, ap);
            }
        }
        for (int i=0; i<n; i++) {
            if (ap[i]) {
                System.out.print(i+" ");
            }
        }
        System.out.println();
    }


    void dfs (int src, int[] disc, int[] low, int[] parent, boolean[] ap) {
        disc[src] = low[src] = ++time;
        int children = 0;
        for (int child: graph.get(src)) {
            if (disc[child] == -1) {
                children++;
                parent[child] = src;
                dfs(child, disc, low, parent, ap);
                low[src] = Math.min(low[src], low[child]);
                if (children > 1 && parent[src] == -1) {
                    ap[src] = true;
                }
                if (parent[src] != -1 && low[child] >= disc[src]) {
                    ap[src] = true;
                }
            }
            else if (parent[src] != child) {
                low[src] = Math.min(low[src], disc[child]);
            }
        }
    }

}


class Main {

	
    public static void main(String[] args) throws IOException {
    	Graph graph = new Graph(5);
        graph.addEdge(0, 1);
        graph.addEdge(0, 2);
        graph.addEdge(1, 2);
        graph.addEdge(2, 3);
        graph.addEdge(3, 4);
        graph.artnPoints();
    }
}

File: 13. Graph/11. Articulation Points/main.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll	long long
#define pb	push_back
#define mp	make_pair
#define fp(i, n)	for (int i = 0; i < (int) n; i++)
#define fpr(i, a, b)	for (int i = (int) a; i <= (int) b; i++)

void fio()
{
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);
}

int timer = 0;
vector<int> points;

void dfs (vector<int> *gr, int src, int par, int *disc, int *low)
{
	disc[src] = low[src] = ++timer;
	int children = 0;
	for (int adj : gr[src]) {
		if (adj == par) {continue;}
		if (disc[adj] != -1) {
			low[src] = min(low[src], disc[adj]);
		}
		else {
			children++;
			dfs(gr, adj, src, disc, low);
			low[src] = min(low[src], low[adj]);
			if (children > 1 && par == -1) {
				points.pb(src);
			}
			if (par != -1 && low[adj] >= disc[src]) {
				points.pb(src);
			}
		}
	}
}

void artnPoints (vector<int> *gr, int n)
{
	int disc[n];
	memset(disc, -1, sizeof(disc));
	int low[n];
	fp(i, n) {
		if (disc[i] == -1) {
			dfs(gr, i, -1, disc, low);
		}
	}
	cout << "Articulation Points:" << "\n";
	for (auto x : points) {
		cout << x << "\n";
	}
}


int main()
{
	fio();

#ifndef ONLINE_JUDGE
	freopen("input.txt", "r", stdin);
	freopen("output.txt", "w", stdout);
#endif

	int n = 5;
	vector<int> gr[n];

	gr[0].pb(1);
	gr[1].pb(0);
	gr[0].pb(2);
	gr[2].pb(0);
	gr[2].pb(1);
	gr[1].pb(2);
	gr[0].pb(3);
	gr[3].pb(0);
	gr[3].pb(4);
	gr[4].pb(3);

	artnPoints(gr, n);

	return 0;
}
